File: swift/Sources/Basecamp/BasecampConfig.swift
import Foundation

/// Configuration for a Basecamp SDK client.
///
/// Provides sensible defaults matching the other SDK implementations.
/// All properties are immutable after construction.
public struct BasecampConfig: Sendable {
    /// Base URL for the Basecamp API.
    public let baseURL: String

    /// User-Agent header value sent with every request.
    public let userAgent: String

    /// Whether to automatically retry on 429/503 responses.
    public let enableRetry: Bool

    /// Whether to enable ETag-based HTTP caching.
    public let enableCache: Bool

    /// Maximum number of pages to follow during pagination (safety cap).
    public let maxPages: Int

    /// Request timeout interval in seconds.
    public let timeoutInterval: TimeInterval

    /// SDK version string.
    public static let version = "0.13.0"

    /// Basecamp API version this SDK targets.
    public static let apiVersion = "2026-08-05"

    /// Default User-Agent header value.
    public static let defaultUserAgent = "basecamp-sdk-swift/\(version) (api:\(apiVersion))"

    /// Default base URL for the Basecamp API.
    public static let defaultBaseURL = "https://3.basecampapi.com"

    /// Creates a new configuration with the given options.
    ///
    /// - Parameters:
    ///   - baseURL: API base URL (default: `https://3.basecampapi.com`)
    ///   - userAgent: User-Agent header (default: `basecamp-sdk-swift/VERSION (api:API_VERSION)`)
    ///   - enableRetry: Enable automatic retry on 429/503 (default: `true`)
    ///   - enableCache: Enable ETag-based caching (default: `false`)
    ///   - maxPages: Maximum pages to follow (default: `10_000`)
    ///   - timeoutInterval: Request timeout in seconds (default: `30`)
    public init(
        baseURL: String = defaultBaseURL,
        userAgent: String = defaultUserAgent,
        enableRetry: Bool = true,
        enableCache: Bool = false,
        maxPages: Int = 10_000,
        timeoutInterval: TimeInterval = 30
    ) {
        // A non-positive cap is not a cap. `BaseService`'s pagination loops are
        // `for _ in 1..<maxPages`, and Swift traps forming a range whose upper
        // bound is below its lower bound — so `maxPages <= 0` already crashes,
        // but deep inside pagination, after page 1 has been fetched, with a
        // message about Range rather than about configuration. Reject it here,
        // at the call that wrote the mistake, the same way the SDK rejects a
        // non-HTTPS base URL and a non-numeric account ID. SPEC.md §2 step 5.
        precondition(maxPages > 0, "maxPages must be positive, got: \(maxPages)")

        self.baseURL = baseURL.hasSuffix("/") ? String(baseURL.dropLast()) : baseURL
        self.userAgent = userAgent
        self.enableRetry = enableRetry
        self.enableCache = enableCache
        self.maxPages = maxPages
        self.timeoutInterval = timeoutInterval
    }
}

File: kotlin/sdk/src/commonMain/kotlin/com/basecamp/sdk/BasecampConfig.kt
package com.basecamp.sdk

import kotlin.time.Duration
import kotlin.time.Duration.Companion.seconds

/**
 * Configuration for a [BasecampClient].
 *
 * Use the builder DSL via [BasecampClient] factory function rather than
 * constructing this directly.
 */
data class BasecampConfig(
    /** Base URL for the Basecamp API. */
    val baseUrl: String = DEFAULT_BASE_URL,
    /** User-Agent header value. */
    val userAgent: String = DEFAULT_USER_AGENT,
    /** Enable ETag-based HTTP caching. */
    val enableCache: Boolean = false,
    /** Enable automatic retry on 429/503 with exponential backoff. */
    val enableRetry: Boolean = true,
    /** Request timeout. */
    val timeout: Duration = DEFAULT_TIMEOUT,
    /** Maximum retry attempts for GET requests. */
    val maxRetries: Int = DEFAULT_MAX_RETRIES,
    /** Maximum pages to follow for pagination (safety cap). */
    val maxPages: Int = DEFAULT_MAX_PAGES,
    /** Base delay for exponential backoff. */
    val baseRetryDelay: Duration = 1.seconds,
) {
    init {
        // A cap of 0 or less is not a cap: BaseService's pagination loops run
        // `while (page < maxPages)`, so a non-positive value consumes zero extra
        // pages and silently returns the first page as if it were the whole
        // collection. Checked here rather than only in BasecampClientBuilder so
        // the invariant travels with the field — the builder constructs this
        // config, so `BasecampClient { maxPages = 0 }` fails through this same
        // require. SPEC.md §2 step 5.
        require(maxPages > 0) {
            "maxPages must be > 0, got: $maxPages"
        }
    }

    companion object {
        const val VERSION = "0.15.0"
        const val API_VERSION = "2026-08-31"
        const val DEFAULT_BASE_URL = "https://3.basecampapi.com"
        const val DEFAULT_USER_AGENT = "basecamp-sdk-kotlin/$VERSION (api:$API_VERSION)"
        const val DEFAULT_MAX_RETRIES = 3
        const val DEFAULT_MAX_PAGES = 10_000
        val DEFAULT_TIMEOUT: Duration = 30.seconds
    }
}
